### PR TITLE
PLANNER-1801: Replace score DRL with constraint provider

### PR DESCRIPTION
**Pull request by yurloc**

### CODE REVIEW

File: optaweb-vehicle-routing-backend/src/main/java/org/optaweb/vehiclerouting/plugin/planner/VehicleRoutingConstraintProvider.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+                .penalizeLong(
+                        "vehicle capacity exceeded",
+                        HardSoftLongScore.ONE_HARD,
+                        (vehicle, demand) -> demand - vehicle.getCapacity()
+                );
+    }
+
+    private static Constraint distanceFromPreviousStandstill(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(PlanningVisit.class)
+                .filter(visit -> visit.getPreviousStandstill() != null)
```
Comment (triceo) [inline optaweb-vehicle-routing-backend/src/main/java/org/optaweb/vehiclerouting/plugin/planner/VehicleRoutingConstraintProvider.java]: Personally, I'd replace the filter with `ifExists(..., Joiner.equals(...))`.
Filters are not indexed, leading to a little drop in performance.
More importantly, I consider that easier to understand than the `null` check.

File: optaweb-vehicle-routing-backend/src/main/java/org/optaweb/vehiclerouting/plugin/planner/VehicleRoutingConstraintProvider.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+                .penalizeLong(
+                        "distance from previous standstill",
+                        HardSoftLongScore.ONE_SOFT,
+                        PlanningVisit::getDistanceFromPreviousStandstill
+                );
+    }
+
+    private static Constraint distanceFromLastVisitToDepot(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(PlanningVisit.class)
+                .filter(visit -> visit.getNextVisit() == null)
```
Comment (triceo) [inline optaweb-vehicle-routing-backend/src/main/java/org/optaweb/vehiclerouting/plugin/planner/VehicleRoutingConstraintProvider.java]: Dtto.

Comment (triceo) [inline optaweb-vehicle-routing-backend/src/main/java/org/optaweb/vehiclerouting/plugin/planner/VehicleRoutingConstraintProvider.java]: Same comment, except `ifNotExists(...)`

File: optaweb-vehicle-routing-backend/src/main/java/org/optaweb/vehiclerouting/plugin/planner/VehicleRoutingConstraintProvider.java (hunk 1 above)
Comment (yurloc) [reply optaweb-vehicle-routing-backend/src/main/java/org/optaweb/vehiclerouting/plugin/planner/VehicleRoutingConstraintProvider.java]: Turns out the filter is not necessary here.

File: optaweb-vehicle-routing-backend/src/main/java/org/optaweb/vehiclerouting/plugin/planner/VehicleRoutingConstraintProvider.java (hunk 2 above)
Comment (yurloc) [reply optaweb-vehicle-routing-backend/src/main/java/org/optaweb/vehiclerouting/plugin/planner/VehicleRoutingConstraintProvider.java:60]: Your comment from another line (for conext):
> Personally, I'd replace the filter with ifExists(..., Joiner.equals(...)).
Filters are not indexed, leading to a little drop in performance.
More importantly, I consider that easier to understand than the null check.

I agree that the null check is not nice. I tried to replace it with `ifNotExists()` without success. I'd really like to see your idea of doing that.

On the other hand, what I'd find even more easy to understand is
```java
.filter(PlanningVisit::isLast)
```
I'm inclined to take that approach for the sake of readability even at some performance cost. What do you think?

Comment (yurloc) [reply optaweb-vehicle-routing-backend/src/main/java/org/optaweb/vehiclerouting/plugin/planner/VehicleRoutingConstraintProvider.java:60]: I'm looking for visits that have no next visit. That means "for visit A, not exists visit B, such that B.previousStandstill equals A". So I tried
```java
return constraintFactory.from(PlanningVisit.class)
    .ifNotExists(PlanningVisit.class, equal(PlanningVisit::getPreviousStandstill))
    // then penalize distance to depot
```
but that always gives me a zero penalty.

Comment (yurloc) [reply optaweb-vehicle-routing-backend/src/main/java/org/optaweb/vehiclerouting/plugin/planner/VehicleRoutingConstraintProvider.java:60]: Solved:
```java
return constraintFactory.from(PlanningVisit.class)
    .ifNotExists(PlanningVisit.class, equal(PlanningVisit::getNextVisit, Function.identity()))
```
`Joiners` needs Javadoc.

File: optaweb-vehicle-routing-backend/pom.xml [hunk 3]
```diff
@@ -130,6 +130,11 @@
       <artifactId>spring-boot-configuration-processor</artifactId>
       <optional>true</optional>
     </dependency>
+    <dependency>
+      <groupId>org.optaplanner</groupId>
+      <artifactId>optaplanner-test</artifactId>
```
Comment (yurloc) [inline optaweb-vehicle-routing-backend/pom.xml:135]: @ge0ffrey This shouldn't be necessary, I think. The dependency should be provided by `optaplanner-spring-boot-starter`.

Comment (ge0ffrey) [reply optaweb-vehicle-routing-backend/pom.xml:135]: Can we do that and still have it scoped as a test dependency?

Comment (ge0ffrey) [reply optaweb-vehicle-routing-backend/pom.xml:135]: Or would we rather just add it to the production dependencies for convenience?

Comment (ge0ffrey) [reply optaweb-vehicle-routing-backend/pom.xml:135]: @triceo ^^

Comment (ge0ffrey) [reply optaweb-vehicle-routing-backend/pom.xml:135]: Added to this issue: https://issues.redhat.com/browse/PLANNER-1957

**Pull request merged**